google/cloud/bigquery/v2/minimal/internal/job_connection.h

```diff
@@ -35,6 +35,7 @@ class BigQueryJobConnection {
   virtual Options options() { return Options{}; }
 
   virtual StatusOr<GetJobResponse> GetJob(GetJobRequest const& request);
+  virtual StatusOr<ListJobsResponse> ListJobs(ListJobsRequest const& request);
```

Comment:
I don't think this is the right API. Part of the value provided by the client library is to take these paginated requests (ones with `next_page_token`) and return something that is iterable. In this repo, `StreamRange<T>` is our iterable type.

I suspect the API should be:
```cc
virtual StreamRange<Job> ListJobs(ListJobsRequest const& request);
```

.... but I am not totally sure, because I am confused about one thing. Is there anything in the `BigqueryHttpResponse` that must be returned in a client call? (Apologies if I missed a previous discussion).

It looks to me like the `FooResponse::BuildFromHttpResponse(BigQueryHttpResponse const& http_response)` methods are intended to extract the relevant information from the `BigQueryHttpResponse`. e.g. in `ListJobs()` we are pulling the `Job` out of the json payload.

https://github.com/googleapis/google-cloud-cpp/blob/c0e30ed9c067609411bdbed5e0440be4cb8e653c/google/cloud/bigquery/v2/minimal/internal/job_response.cc#L54-L68

When I look at what else is in the type, there is an HTTP status code and headers.

https://github.com/googleapis/google-cloud-cpp/blob/c0e30ed9c067609411bdbed5e0440be4cb8e653c/google/cloud/bigquery/v2/minimal/internal/bigquery_http_response.h#L34-L36

It's my understanding that we do some translation of HTTP status code -> `google::cloud::StatusCode`. And that information get's captured in the `Status` part of the `StatusOr<>` that is returned.

https://github.com/googleapis/google-cloud-cpp/blob/c0e30ed9c067609411bdbed5e0440be4cb8e653c/google/cloud/internal/rest_response.h#L84

So that just leaves the headers, and I am not sure what they are used for. Do they need to be returned to the caller?

---

Maybe you based this off of something like `ListBuckets` in `storage`. I suspect if we were writing it from scratch, we would choose to return an iterable type instead of something with a page token.

https://github.com/googleapis/google-cloud-cpp/blob/c0e30ed9c067609411bdbed5e0440be4cb8e653c/google/cloud/storage/internal/bucket_requests.h#L58-L66

https://github.com/googleapis/google-cloud-cpp/blob/c0e30ed9c067609411bdbed5e0440be4cb8e653c/google/cloud/storage/internal/rest_client.h#L60-L61

---

If I am right, I suspect the other API should be:
```cc
virtual StatusOr<Job> GetJob(GetJobRequest const& request);
```